connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -0,0 +1,169 @@
+ *    started on another worker, we may not have the updated status cached yet. In this case, we'll overwrite
+ *    the value which will cause the state to be inconsistent (most likely until the next rebalance). Until
+ *    we have proper producer groups with fenced groups, there is not much else we can do.
+ */
+public abstract class AbstractHerder implements Herder, TaskStatus.Listener, ConnectorStatus.Listener {
+
+    protected final StatusBackingStore statusBackingStore;
+    private final String workerId;
+
+    public AbstractHerder(StatusBackingStore statusBackingStore, String workerId) {
+        this.statusBackingStore = statusBackingStore;
+        this.workerId = workerId;
+    }
+
+    protected abstract int generation();
+
+    protected void startServices() {
+        this.statusBackingStore.start();
+    }
+
+    protected void stopServices() {
+        this.statusBackingStore.stop();
+    }
+
+    @Override
+    public void onStartup(String connector) {
+        statusBackingStore.put(new ConnectorStatus(connector, ConnectorStatus.State.RUNNING,
+                workerId, generation()));
+    }
+
+    @Override
+    public void onShutdown(String connector) {
+        statusBackingStore.putSafe(new ConnectorStatus(connector, ConnectorStatus.State.UNASSIGNED,
+                workerId, generation()));
+    }
+
+    @Override
+    public void onFailure(String connector, Throwable cause) {
+        statusBackingStore.putSafe(new ConnectorStatus(connector, ConnectorStatus.State.FAILED,
+                message(cause), workerId, generation()));
+    }
+
+    @Override
+    public void onStartup(ConnectorTaskId id) {
+        statusBackingStore.put(new TaskStatus(id, TaskStatus.State.RUNNING, workerId, generation()));
+    }
+
+    @Override
+    public void onFailure(ConnectorTaskId id, Throwable cause) {
+        statusBackingStore.putSafe(new TaskStatus(id, TaskStatus.State.FAILED, message(cause), workerId, generation()));
+    }
+
+    @Override
+    public void onShutdown(ConnectorTaskId id) {
+        statusBackingStore.putSafe(new TaskStatus(id, TaskStatus.State.UNASSIGNED, workerId, generation()));
+    }
+
+    @Override
+    public void onDeletion(String connector) {
+        for (TaskStatus status : statusBackingStore.getAll(connector))
+            statusBackingStore.put(new TaskStatus(status.id(), TaskStatus.State.DESTROYED, workerId, generation()));
+        statusBackingStore.put(new ConnectorStatus(connector, ConnectorStatus.State.DESTROYED, workerId, generation()));
+    }
+
+    @Override
+    public void connectorStatus(String connName, Callback<ConnectorStateInfo> callback) {
+        ConnectorStateInfo status = connectorStatus(connName);
+
+        if (status == null)
+            throw new NotFoundException("No status found for connector " + connName);
+        else
+            callback.onCompletion(null, status);
```

Comment:
Why a callback here? Just for uniformity? It seems like we're not ever doing anything async so there doesn't seem to be a need for it. Same re: taskStatus below.

On the other hand, is there any issue during startup when reading to log end offset of the status topic where we might want it to be async?